Error: 1466 - must not specify _key for this collection

This started happening on cluster deployments on attempts to insert data after #1048 fix. The error is apparently related to the `dataSource` collection although the error message isn't specific about it.

```
java.util.concurrent.CompletionException: com.arangodb.ArangoDBException: Response: 400, Error: 1466 - must not specify _key for this collection
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367)
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:376)
	at java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1019)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at com.arangodb.async.internal.velocystream.VstCommunicationAsync.lambda$execute$1(VstCommunicationAsync.java:101)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at com.arangodb.async.internal.velocystream.VstConnectionAsync.lambda$write$0(VstConnectionAsync.java:51)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.arangodb.internal.velocystream.internal.MessageStore.consume(MessageStore.java:62)
	at com.arangodb.internal.velocystream.internal.ChunkStore.checkCompleteness(ChunkStore.java:65)
	at com.arangodb.internal.velocystream.internal.ChunkStore.checkCompleteness(ChunkStore.java:59)
	at com.arangodb.internal.velocystream.internal.VstConnection.lambda$open$0(VstConnection.java:218)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.arangodb.ArangoDBException: Response: 400, Error: 1466 - must not specify _key for this collection
	at com.arangodb.internal.util.ResponseUtils.checkError(ResponseUtils.java:55)
	at com.arangodb.internal.velocystream.VstCommunication.checkError(VstCommunication.java:157)
	at com.arangodb.async.internal.velocystream.VstCommunicationAsync.lambda$execute$1(VstCommunicationAsync.java:80)
	... 14 common frames omitted
```

fixes #1076

--- a/persistence/src/main/scala/za/co/absa/spline/persistence/ArangoManager.scala
+++ b/persistence/src/main/scala/za/co/absa/spline/persistence/ArangoManager.scala
@@ -67,7 +67,7 @@ class ArangoManagerImpl(
         Future.successful(false)
       } else for {
         _ <- deleteDbIfRequested(db, onExistsAction == Drop)
-        _ <- db.create().toScala
+        _ <- createDb(db)
         _ <- createCollections(db, options)
         _ <- createAQLUserFunctions(db)
         _ <- createFoxxServices()
@@ -136,6 +136,11 @@ class ArangoManagerImpl(
       }
       else Future.successful({})
     } yield {}
+  }
+
+  private def createDb(db: ArangoDatabaseAsync) = {
+    log.info(s"Create database: ${db.dbName}")
+    db.create().toScala
   }
 
   private def createCollections(db: ArangoDatabaseAsync, options: DatabaseCreateOptions) = {

--- a/persistence/src/main/scala/za/co/absa/spline/persistence/model/entities.scala
+++ b/persistence/src/main/scala/za/co/absa/spline/persistence/model/entities.scala
@@ -86,10 +86,11 @@ object DBVersion {
 case class DataSource(
   uri: DataSource.Uri,
   name: DataSource.Name,
-  override val _key: DataSource.Key,
   lastWriteDetails: Option[Progress]
 ) extends Vertex with RootEntity {
-  def this() = this(null, null, null, null)
+  override val _key: DataSource.Key = null
+
+  def this() = this(null, null, null)
 }
 
 object DataSource {
@@ -99,9 +100,11 @@ object DataSource {
 
   private val NameRegexp = "([^/]+)/*$".r
 
-  def getName(uri: Uri): Name = {
+  private[persistence] def getName(uri: Uri): Name = {
     NameRegexp.findFirstMatchIn(uri).map(_.group(1)).getOrElse("")
   }
+
+  def apply(uri: Uri): DataSource = DataSource(uri, getName(uri), None)
 }
 
 /**

--- a/producer-services/src/main/scala/za/co/absa/spline/producer/service/model/ExecutionPlanPersistentModel.scala
+++ b/producer-services/src/main/scala/za/co/absa/spline/producer/service/model/ExecutionPlanPersistentModel.scala
@@ -35,9 +35,6 @@ case class ExecutionPlanPersistentModel(
   uses: Seq[Edge], // ... attribute or expression
   produces: Seq[Edge], // ... attribute
 
-  // data source
-  dataSources: Seq[DataSource],
-
   // schema
   schemas: Seq[Schema],
   consistsOf: Seq[Edge], // ... attributes

--- a/producer-services/src/main/scala/za/co/absa/spline/producer/service/model/ExecutionPlanPersistentModelBuilder.scala
+++ b/producer-services/src/main/scala/za/co/absa/spline/producer/service/model/ExecutionPlanPersistentModelBuilder.scala
@@ -31,11 +31,9 @@ import za.co.absa.spline.producer.model.{v1_2 => am}
 import za.co.absa.spline.producer.service.model.ExecutionPlanPersistentModelBuilder._
 import za.co.absa.spline.producer.service.{InconsistentEntityException, model}
 
-import java.util.UUID.randomUUID
-
 class ExecutionPlanPersistentModelBuilder private(
   ep: am.ExecutionPlan,
-  persistedDSKeyByURI: Map[pm.DataSource.Uri, pm.DataSource.Key]
+  dsKeyByURI: pm.DataSource.Uri => pm.DataSource.Key
 ) {
   private val keyCreator = new ExecutionPlanKeyCreator(ep)
   private val epPKey: pm.ArangoDocument.Id = ep.id.toString
@@ -63,18 +61,6 @@ class ExecutionPlanPersistentModelBuilder private(
   // attr dependency graph
   private var _attrDepGraph = Graph.empty[AttrOrExprRef, DiEdge]
 
-  private val pmDataSourceByURI: Map[String, pm.DataSource] = {
-    val referencedURIs = ep.dataSources
-    val persistedURIs = persistedDSKeyByURI.keys
-    val transientDSKeyByURI = (referencedURIs -- persistedURIs).map(_ -> randomUUID.toString).toMap
-    val dsKeyByUri = transientDSKeyByURI ++ persistedDSKeyByURI
-    dsKeyByUri.map {
-      case (uri, key) =>
-        val dsName = pm.DataSource.getName(uri)
-        uri -> pm.DataSource(uri, dsName, key, None)
-    }
-  }
-
   def build(): ExecutionPlanPersistentModel = {
     val pmExecutionPlan = pm.ExecutionPlan(
       name = ep.name,
@@ -100,11 +86,6 @@ class ExecutionPlanPersistentModelBuilder private(
         )
       }
 
-    val pmTransientDataSources = {
-      val persistentDSUris = persistedDSKeyByURI.keys
-      (pmDataSourceByURI -- persistentDSUris).values.toSeq
-    }
-
     model.ExecutionPlanPersistentModel(
       // plan
       executionPlan = pmExecutionPlan,
@@ -120,9 +101,6 @@ class ExecutionPlanPersistentModelBuilder private(
       emits = _pmEmits,
       uses = _pmUses,
       produces = _pmProduces,
-
-      // data source
-      dataSources = pmTransientDataSources,
 
       // schema
       schemas = _pmSchemas,
@@ -257,7 +235,7 @@ class ExecutionPlanPersistentModelBuilder private(
     } yield {
       EdgeDef.ReadsFrom.edge(
         keyCreator.asOperationKey(ro.id),
-        pmDataSourceByURI(ds)._key,
+        dsKeyByURI(ds),
         epPKey
       )
     }
@@ -265,7 +243,7 @@ class ExecutionPlanPersistentModelBuilder private(
   private def pmWritesTo: pm.Edge = {
     EdgeDef.WritesTo.edge(
       keyCreator.asOperationKey(ep.operations.write.id),
-      pmDataSourceByURI(ep.operations.write.outputSource)._key,
+      dsKeyByURI(ep.operations.write.outputSource),
       epPKey
     )
   }
@@ -276,14 +254,14 @@ class ExecutionPlanPersistentModelBuilder private(
       ds <- ro.inputSources
     } yield EdgeDef.Depends.edge(
       ep.id,
-      pmDataSourceByURI(ds)._key,
+      dsKeyByURI(ds),
       epPKey
     )
 
   private def pmAffects: pm.Edge =
     EdgeDef.Affects.edge(
       ep.id,
-      pmDataSourceByURI(ep.operations.write.outputSource)._key,
+      dsKeyByURI(ep.operations.write.outputSource),
       epPKey
     )
 
@@ -346,10 +324,11 @@ object ExecutionPlanPersistentModelBuilder {
 
   def toPersistentModel(
     ep: am.ExecutionPlan,
-    persistedDSKeyByURI: Map[pm.DataSource.Uri, pm.DataSource.Key]
+    persistedDataSources: Seq[pm.DataSource]
   ): ExecutionPlanPersistentModel = {
     val maybeExpressions = ep.expressions.map(_.all)
-    new ExecutionPlanPersistentModelBuilder(ep, persistedDSKeyByURI)
+    val dsKeyByURI = persistedDataSources.map(ds => ds.uri -> ds._key).toMap
+    new ExecutionPlanPersistentModelBuilder(ep, dsKeyByURI)
       .addOperations(ep.operations.all)
       .addAttributes(ep.attributes)
       .having(maybeExpressions)(_ addExpressions _)

--- a/producer-services/src/main/scala/za/co/absa/spline/producer/service/repo/ExecutionProducerRepositoryImpl.scala
+++ b/producer-services/src/main/scala/za/co/absa/spline/producer/service/repo/ExecutionProducerRepositoryImpl.scala
@@ -29,7 +29,6 @@ import za.co.absa.spline.producer.service.UUIDCollisionDetectedException
 import za.co.absa.spline.producer.service.model.{ExecutionEventKeyCreator, ExecutionPlanPersistentModel, ExecutionPlanPersistentModelBuilder}
 
 import scala.compat.java8.FutureConverters._
-import scala.compat.java8.StreamConverters._
 import scala.concurrent.{ExecutionContext, Future}
 import scala.util.control.NonFatal
 
@@ -52,18 +51,23 @@ class ExecutionProducerRepositoryImpl @Autowired()(db: ArangoDatabaseAsync) exte
       Map("key" -> executionPlan.id)
     )
 
-    val eventualPersistedDSKeyByURI: Future[Map[DataSource.Uri, DataSource.Key]] = db.queryAs[DataSource](
-      s"""
-         |WITH ${NodeDef.DataSource.name}
-         |FOR ds IN ${NodeDef.DataSource.name}
-         |    FILTER ds.uri IN @refURIs
-         |    RETURN KEEP(ds, ['_key', 'uri'])
-         |    """.stripMargin,
-      Map("refURIs" -> executionPlan.dataSources.toArray)
-    ).map(_.streamRemaining.toScala.map(ds => ds.uri -> ds._key).toMap)
+    val eventualPersistedDataSources: Future[Seq[DataSource]] = {
+      val dataSources: Set[DataSource] = executionPlan.dataSources.map(DataSource.apply)
+      db.queryStream[DataSource](
+        s"""
+           |WITH ${NodeDef.DataSource.name}
+           |FOR ds IN @dataSources
+           |    UPSERT { uri: ds.uri }
+           |        INSERT KEEP(ds, ['_created', 'uri', 'name'])
+           |        UPDATE {} IN ${NodeDef.DataSource.name}
+           |        RETURN KEEP(NEW, ['_key', 'uri'])
+           |    """.stripMargin,
+        Map("dataSources" -> dataSources.toArray)
+      )
+    }
 
     for {
-      persistedDSKeyByURI <- eventualPersistedDSKeyByURI
+      persistedDSKeyByURI <- eventualPersistedDataSources
       maybeExistingDiscriminatorOpt <- eventualMaybeExistingDiscriminatorOpt
       _ <- maybeExistingDiscriminatorOpt match {
         case Some(existingDiscriminatorOrNull) =>
@@ -100,10 +104,10 @@ object ExecutionProducerRepositoryImpl {
 
   private def createInsertTransaction(
     executionPlan: apiModel.ExecutionPlan,
-    persistedDSKeyByURI: Map[DataSource.Uri, DataSource.Key]
+    persistedDataSources: Seq[DataSource]
   ) = {
     val eppm: ExecutionPlanPersistentModel =
-      ExecutionPlanPersistentModelBuilder.toPersistentModel(executionPlan, persistedDSKeyByURI)
+      ExecutionPlanPersistentModelBuilder.toPersistentModel(executionPlan, persistedDataSources)
 
     new TxBuilder()
       // execution plan
@@ -120,9 +124,6 @@ object ExecutionProducerRepositoryImpl {
       .addQuery(InsertQuery(EdgeDef.Emits, eppm.emits))
       .addQuery(InsertQuery(EdgeDef.Uses, eppm.uses))
       .addQuery(InsertQuery(EdgeDef.Produces, eppm.produces))
-
-      // data source
-      .addQuery(InsertQuery(NodeDef.DataSource, eppm.dataSources))
 
       // schema
       .addQuery(InsertQuery(NodeDef.Schema, eppm.schemas))

